core/src/main/scala/org/bitcoins/core/util/NumberUtil.scala

```diff
@@ -175,19 +177,121 @@ trait NumberUtil extends BitcoinSLogger {
+
+        val pow256 = BigInteger.valueOf(256).pow(exponent)
+        mantissa.multiply(pow256)
+      }
+    }
+
+    val nWordNotZero = mantissa != (BigInteger.ZERO)
+    val nWord = BigInt(mantissa)
+    val nSize = nBits.toBigInt / (NumberUtil.pow2(24))
+    val isNegative: Boolean = {
+      //*pfNegative = nWord != 0 && (nCompact & 0x00800000) != 0;
+      nWordNotZero &&
+      (nBits & UInt32.fromHex("00800000")) != UInt32.zero
+    }
+
+    val isOverflow: Boolean = {
+      //nWord != 0 && ((nSize > 34) ||
+      //  (nWord > 0xff && nSize > 33) ||
+      //  (nWord > 0xffff && nSize > 32));
+
+      nWordNotZero && ((nSize > 34) ||
+      (nWord > UInt8.max.toBigInt && nSize > 33) ||
+      (nWord > UInt32.fromHex("ffff").toBigInt && nSize > 32))
+    }
+
+    BlockHeader.TargetDifficultyHelper(result.abs(), isNegative, isOverflow)
+  }
+
+  /**
+    * Compressed the big integer to be used inside of [[org.bitcoins.core.protocol.blockchain.BlockHeader.nBits]]
+    * [[https://github.com/bitcoin/bitcoin/blob/2068f089c8b7b90eb4557d3f67ea0f0ed2059a23/src/arith_uint256.cpp#L226 bitcoin core implementation]]
+    * @param bigInteger
+    * @return
+    */
+  def targetCompression(bigInteger: BigInteger, isNegative: Boolean): UInt32 = {
+    val bytes = bigInteger.toByteArray
+    val bitVec = BitVector(bytes)
+
+    val negativeFlag = UInt32.fromHex("00800000")
+
+    //emulates bits() in arith_uin256.h
+    //Returns the position of the highest bit set plus one, or zero if the
+    //value is zero.
+    //https://github.com/bitcoin/bitcoin/blob/2068f089c8b7b90eb4557d3f67ea0f0ed2059a23/src/arith_uint256.h#L241
+    var size: Int = if (bigInteger == BigInteger.valueOf(0)) {
+      0
     } else {
-      val exponent = significand - 3
+      ((bitVec.length + 7) / 8).toInt
+    }
+
+    var compact: UInt32 = {
+      if (size <= 3) {
+        // GetLow64() << 8 * (3 - nSize);
+        //note: counter intuitively, the expression
+        // 8 * (3 - nSize)
+        //gets evaluated before we apply the shift left operator
+        //verified this property with g++
+        val shiftAmount = 8 * (3 - size)
 
-      val pow256 = BigInteger.valueOf(256).pow(exponent)
-      mantissa.multiply(pow256)
+        val u64 = new BigInteger(1, bitVec.takeRight(64).toByteArray)
+          .shiftLeft(shiftAmount)
+
+        UInt32.fromBytes(ByteVector(u64.toByteArray))
+      } else {
+        //8 * (nSize - 3)
+        val shiftAmount = 8 * (size - 3)
+        val bn = bigInteger.shiftRight(shiftAmount)
+        val bytes = bn.toByteArray.takeRight(4)
+        UInt32.fromBytes(ByteVector(bytes))
+      }
     }
 
+    //The 0x00800000 bit denotes the sign.
+    // Thus, if it is already set, divide the mantissa by 256 and increase the exponent.
+    if ((compact & negativeFlag) != UInt32.zero) {
+      compact = compact >> 8
+      size = size + 1
+    }
+
+    //~0x007fffff = 0xff800000
+    require((compact & UInt32.fromHex("ff800000")) == UInt32.zero)
```

Comment:
`UInt32(0xff800000)`?

Reply:
This does have a problem, since using the raw number construction `0x` that scala provides, and the first bit is a `1`, you end up with a negative number which can't be used with a `UInt32`

```scala
scala> val u32 = UInt32(0x80000000)
<console>:10: warning: Unused import
       import org.bitcoins.core.number._
                                       ^
java.lang.IllegalArgumentException: requirement failed: We cannot have a negative number in an unsigned number, got: -2147483648
  at scala.Predef$.require(Predef.scala:281)
  at org.bitcoins.core.number.UInt32$UInt32Impl.<init>(NumberType.scala:300)
  at org.bitcoins.core.number.UInt32$.apply(NumberType.scala:322)
  at org.bitcoins.core.number.UInt32$.apply(NumberType.scala:320)
  ... 36 elided
```

Reply:
`UInt32(0xff800000L)` solves it